File: Data Structures & Algorithms/three-integer-sum/submission-2.py
#nums[i] = -(nums[j] + nums[k])
# -nums[i] = nums[j] + nums[k]

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = set()
        nums.sort()


        for i in range(len(nums)):
            left = i + 1
            right = len(nums) - 1
            target = -(nums[i])

            while left < right:
                if nums[left] + nums[right] < target:
                    left += 1
                    continue
                if nums[left] + nums[right] > target:
                    right -= 1
                    continue
                if nums[left] + nums[right] == target:
                    result.add(tuple([nums[i], nums[left], nums[right]]))
                    left += 1

            

            
        return list(result)
                
                

File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */

    //can k be 0? can the input array be empty
    // we can track frequencies using a hashmap
    // then we can get the entries, sort by values, and return the top k keys
    
    topKFrequent(nums, k) {
        let count = {}
        let result = []
        for (let i = 0; i < nums.length; i++){
            if (nums[i] in count){
                count[nums[i]] += 1
            }
            else{
                count[nums[i]] = 1
            }
        }

        let counts = Object.entries(count).sort((a, b) => b[1]-a[1]) //descending order
        for (let i = 0; i < k; i++){
            result.push(counts[i][0])
        }

        return result
    }
}
